Implement a streaming serializer for watch

smarterclayton opened the pull request:
Changeover watch to use streaming serialization. Properly version the
watch objects. Implement simple framing for JSON and Protobuf (but not
YAML).

@wojtek-t @lavalamp

smarterclayton commented:
Not complete yet, I'm seeing round trip serialization failures on resources bound into multiple groups (Scale is serializing the wrong thing in "autoscaling" but the right thing in "extensions").

wojtek-t commented on cmd/libs/go2idl/client-gen/testdata/apis/testgroup/v1/register.go:
```diff
@@ -41,6 +42,7 @@ func addKnownTypes(scheme *runtime.Scheme) {
 		&v1.DeleteOptions{},
 		&unversioned.Status{},
 		&v1.ExportOptions{})
+	versionedwatch.AddToGroupVersion(scheme, SchemeGroupVersion)
```
I don't understand why we need to add it here - it seems to be that your changes doesn't affect client-generation code.

smarterclayton replied:
This is defining an API group, that is used by `examples/apiserver/apiserver.go` to serve API requests.  To serve api requests you need to have the WatchEvent API type defined now.

wojtek-t replied:
got it

wojtek-t commented on cmd/libs/go2idl/go-to-protobuf/protobuf/cmd.go:
```diff
@@ -64,6 +64,7 @@ func New() *Generator {
 			`+k8s.io/kubernetes/pkg/api/resource`,
 			`+k8s.io/kubernetes/pkg/runtime`,
 			`k8s.io/kubernetes/pkg/api/unversioned`,
+			`+k8s.io/kubernetes/pkg/watch/versioned`,
```
nit: can you please move this line one above (so that we have all "+" lines together)?

wojtek-t commented on pkg/api/serialization_proto_test.go:
```diff
@@ -38,7 +38,7 @@ import (
 func init() {
 	codecsToTest = append(codecsToTest, func(version unversioned.GroupVersion, item runtime.Object) (runtime.Codec, error) {
-		s := protobuf.NewSerializer(api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
+		s := protobuf.NewSerializer(api.Scheme, runtime.ObjectTyperToTyper(api.Scheme), "application/vnd.kubernetes.protobuf")
```
Since this is main proto-serialization-related test, can you please use this ContentTypeProtobuf constant instead (I know it's not that important in tests, but it would make me feel more comfortable :))

wojtek-t commented on pkg/api/serialization_test.go:
```diff
@@ -68,13 +70,18 @@ func roundTrip(t *testing.T, codec runtime.Codec, item runtime.Object) {
 	}
 
+	dataString := string(data)
+	if !strings.HasPrefix(dataString, "{") {
```
Can you please explain why this "if" is needed?

smarterclayton replied:
To support both JSON and protobuf without inventing a new abstraction?

wojtek-t replied:
Sure - but for proto you have this "serialization_proto_test", right? So why changing it here?

smarterclayton replied:
serialization_proto_test just injects a new dimension to fuzz test (JSON + protobuf)

wojtek-t commented on pkg/apiserver/apiserver_test.go:
```diff
@@ -176,6 +178,8 @@ func addNewTestTypes() {
 		&api.DeleteOptions{}, &apiservertesting.SimpleGetOptions{}, &apiservertesting.SimpleRoot{})
 	api.Scheme.AddKnownTypes(newGroupVersion, &v1.Pod{})
+	versioned.AddToGroupVersion(api.Scheme, newGroupVersion)
+	api.Scheme.AddKnownTypes(newGroupVersion, &v1.Pod{})
```
please remove - this is duplicated from 2 lines above

xiang90 commented:
/cc @hongchaodeng

wojtek-t commented on pkg/apiserver/watch.go:
```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+	}
+
+	// find the embedded serializer matching the media type
+	// TODO: store content type on runtime.Unknown and serialize that into the stream, allowing for mixed payloads
```
runtime.Unknown already has ContentType, right?

smarterclayton replied:
Yeah but extension does not, so the TODO we left in your PR needs to be
part of this PR I think.

On Mon, Apr 4, 2016 at 12:40 PM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/apiserver/watch.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r58406427>
> :
>
> >  	if err != nil {
> >  		scope.err(err, res.ResponseWriter, req.Request)
> >  		return
> >  	}
> > -	// TODO: replace with typed serialization
> > -	if mediaType != "application/json" {
> > -		writeRawJSON(http.StatusNotAcceptable, (errNotAcceptable{[]string{"application/json"}}).Status(), res.ResponseWriter)
> > +	encoder := scope.StreamSerializer.EncoderForVersion(serializer, scope.Kind.GroupVersion())
> > +
> > +	useTextFraming := false
> > +	if encodable, ok := encoder.(textEncodable); ok && encodable.EncodesAsText() {
> > +		useTextFraming = true
> > +	}
> > +
> > +	// find the embedded serializer matching the media type
> > +	// TODO: store content type on runtime.Unknown and serialize that into the stream, allowing for mixed payloads
>
> runtime.Unknown already has ContentType, right?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/26e6630171ddc8f3653bbcfa270efb87e1073818#r58406427>
>

wojtek-t replied:
You mean - RawExtension, right? I don't understand how it is related - can you please clarify?

smarterclayton replied:
Hrm - I was thinking this was the other section in the code.  This todo can be removed.

smarterclayton replied:
The protobuf serializer probably needs to be able to encode anything it can't recognize as a "nested" type (or just arbitrarily by being passed a parent JSON serializer).  Until we sort through what that looks like (which isn't implemented in this PR, but we do have a TODO there), it's possible that the embeddedSerializer might need to decide what to do with fallback cases.

wojtek-t commented on pkg/apiserver/watch.go:
```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
+				return
+			}
+			event.Object = &runtime.Unknown{
```
So that means that we are always sending "runtim.Unknown" as part of our watch.
I'm not saying it's bad, but maybe we should document it somewhere.

@lavalamp - FYI ^^

smarterclayton replied:
No, the serialization of runtime.Unknown is unwrapped to send the actual
body for JSON and is the normal serialization for protobuf.

On Mon, Apr 4, 2016 at 1:01 PM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In pkg/apiserver/watch.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r58409606>
> :
>
> >  			if !ok {
> >  				// End of results.
> >  				return
> >  			}
> > -			self.fixup(event.Object)
> > -			if err := encoder.Encode(&event); err != nil {
> > -				// Client disconnect.
> > +			obj := event.Object
> > +			s.fixup(obj)
> > +			if err := s.embeddedEncoder.EncodeToStream(obj, buf); err != nil {
> > +				// unexpected error
> > +				utilruntime.HandleError(fmt.Errorf("unable to encode watch object: %v", err))
> > +				return
> > +			}
> > +			event.Object = &runtime.Unknown{
>
> So that means that we are always sending "runtim.Unknown" as part of our
> watch.
> I'm not saying it's bad, but maybe we should document it somewhere.
>
> @lavalamp <https://github.com/lavalamp> - FYI ^^
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/26e6630171ddc8f3653bbcfa270efb87e1073818#r58409606>
>

wojtek-t replied:
Sorry - I'm probably missing something. So the "Encode" method in case of runtime.Unknown is not serializating the Unknown object, but is just taking the Raw field and returns it? Is that the case?

smarterclayton replied:
Each serializer handles runtime.Unknown being passed to it.  The behavior should roughly be - if you get a runtime.Unknown without a content type, assume the content type is the same as the object and encode the raw bytes.  However, since some protobuf output needs to be wrapped with type info, the Serializer and RawSerializer for protobuf behave differently.  The serializer would write the runtime.Unknown itself (with type info added).  The raw serializer would write the byte contents only.

By passing runtime.Unknown here we're saying "we've already encoded this object".  The embedded encoder *must* add type info (since the event is a true generic runtime.Unknown).  The stream encoder is a raw serializer, and only writes the raw event protobuf fields (since events aren't typed themselves, don't have TypeMeta)

wojtek-t replied:
Thanks for explanation,

wojtek-t commented on pkg/apiserver/watch.go:
```diff
@@ -64,119 +59,218 @@ func (w *realTimeoutFactory) TimeoutCh() (<-chan time.Time, func() bool) {
+				// ContentType is not required here because we are defaulting to the serializer
+				// type
+			}
+			internalEvent := versioned.InternalEvent(event)
```
In general - shouldn't we do the conversion here? Maybe add a TODO about it? WDYT?

smarterclayton replied:
The encoder is responsible for understanding how to convert an internal event to the proper format and version.  I can add a comment to that effect.

wojtek-t commented on pkg/api/serialization_proto_test.go:
```diff
@@ -38,7 +38,7 @@ import (
 func init() {
 	codecsToTest = append(codecsToTest, func(version unversioned.GroupVersion, item runtime.Object) (runtime.Codec, error) {
-		s := protobuf.NewSerializer(api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
+		s := protobuf.NewSerializer(api.Scheme, runtime.ObjectTyperToTyper(api.Scheme), "application/arbitrary.content.type")
```
Since this is main proto-serialization-related test, can you please use this ContentTypeProtobuf constant instead (I know it's not that important in tests, but it would make me feel more comfortable :))

smarterclayton replied:
My concern is that it *shouldn't* affect any of the serialization - it should only impact whatever the default value selected is in the nesting.  My worry is that mime type is a very special value that needs to be kept located in very specific places, not littered throughout this code.  Serialization doesn't depend on mime types, only nested encoding does (and nested encoding should probably be via a nested encoder that would specify a specific type).

wojtek-t replied:
OK - that makes sense. Thanks for explanation.

smarterclayton replied:
Yeah, I'm digging into why the general serialization is not roundtripping,
hopefully have that fixed tomorrow

On Mon, Apr 4, 2016 at 4:05 PM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In pkg/api/serialization_proto_test.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r58438663>
> :
>
> > @@ -38,7 +38,7 @@ import (
> >
> >  func init() {
> >  	codecsToTest = append(codecsToTest, func(version unversioned.GroupVersion, item runtime.Object) (runtime.Codec, error) {
> > -		s := protobuf.NewSerializer(api.Scheme, runtime.ObjectTyperToTyper(api.Scheme))
> > +		s := protobuf.NewSerializer(api.Scheme, runtime.ObjectTyperToTyper(api.Scheme), "application/arbitrary.content.type")
>
> OK - that makes sense. Thanks for explanation.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/828d12e0f26fa80d52b22cc46d73f2f883b3b403#r58438663>
>

wojtek-t commented on pkg/registry/thirdpartyresourcedata/codec.go:
```diff
@@ -432,7 +432,8 @@ func (t *thirdPartyResourceDataCreator) New(kind unversioned.GroupVersionKind) (
 		}
 		return &extensions.ThirdPartyResourceDataList{}, nil
-	case "ListOptions":
+		// TODO: this list needs to be formalized higher in the chain
```
nit: can you un-indent it by one tab?

wojtek-t commented on pkg/util/framer/framer.go:
```diff
@@ -0,0 +1,121 @@
+// ReadFrame decodes the next JSON object in the stream, or returns an error. The returned
+// byte slice will be modified the next time ReadFrame is invoked and should not be altered.
+func (r *jsonFrameReader) Read(data []byte) (int, error) {
+	m := json.RawMessage(data[:0])
```
I don't understand this - shouldn't it be data[0:] ?
If not - can you please add a comment?

wojtek-t commented:
@smarterclayton - thanks for all explanations! Can you please fix the compile errors?

smarterclayton commented:
Whew... so the good news is that I figured out what was broken, the bad news was that it was gogoproto, and I'm probably the one who did it :)

wojtek-t commented:
:) - can you roughly explain what exactly it was?

smarterclayton commented:
As soon as I figure it out :)

lavalamp commented:
Haven't had a chance to look at this yet-- is this cross version
compatible? (does it change the current output of watch?)

On Mon, Apr 4, 2016 at 4:39 PM, Clayton Coleman <notifications@github.com>
wrote:

> As soon as I figure it out :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806#issuecomment-205543132>
>

wojtek-t commented:
@lavalamp - it shouldn't - the goal is to have this backward compatible

smarterclayton commented:
Yeah, this should be 100% byte for byte *EXCEPT* watch is now delimited by
newline, rather than just by }{.  That's because we were not using the json
encoder before to frame our responses.  It also sets up future conversion
between versions.

Also, debugging binary protocols is not fun.

On Mon, Apr 4, 2016 at 8:35 PM, Wojciech Tyczynski <notifications@github.com
> wrote:

> @lavalamp <https://github.com/lavalamp> - it shouldn't - the goal is to
> have this backward compatible
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806#issuecomment-205560527>
>

smarterclayton committed:
Implement correct and memory safe framing and add tests

smarterclayton committed:
Fix compilation errors in protobuf and add decode test

smarterclayton committed:
Implement a streaming serializer for watch

Changeover watch to use streaming serialization. Properly version the
watch objects. Implement simple framing for JSON and Protobuf (but not
YAML).

smarterclayton commented:
Problem was passing the wrong data down into the serializer.  Which took about 5 hours to debug...

PR is updated with commits split.  I simplified Timestamp serialization (first two commits).  I updated and added tests for the framers and the streaming serializer, and fixed the compile errors in the protobuf serializer and added a decode test.

smarterclayton commented:
Round trips are passing for all objects, which is a good sign (except for those that the defaulter is hiding).

wojtek-t commented on pkg/api/unversioned/time_proto.go:
```diff
@@ -22,37 +22,23 @@ import (
-	// that count forward in time. Must be from 0 to 999,999,999
-	// inclusive.
-	Nanos int32 `json:"nanos"`
+	Seconds int64 `json:"seconds" protobuf:"varint,1,opt,name=seconds"`
```
hmm - why you are removing nanos?

smarterclayton replied:
We don't support nanos on any of our time fields today (they're zeroed) but I can add them back here.  We'd need to set our expectations on time which we've never done.

wojtek-t commented:
@smarterclayton - will do my best to review it later today

One comment - can you separate the first two commits into a separate PR? Those are kind of independent.

smarterclayton commented:
Yeah I can split

smarterclayton commented:
Split first two commits into #23871

smarterclayton commented:
@k8s-bot test this: #23878

wojtek-t commented on pkg/api/serialization_test.go:
```diff
@@ -58,6 +61,15 @@ func fuzzInternalObject(t *testing.T, forVersion unversioned.GroupVersion, item
+	dataString := string(data)
+	if !strings.HasPrefix(dataString, "{") {
+		dataString = "\n" + hex.Dump(data)
+		proto.NewBuffer(make([]byte, 0, 1024)).DebugPrint("decoded object", data)
```
Just curious - what's that?

smarterclayton replied:
It's a helper that shows using proto rules the structs and their types.
It's moderately useful when debugging proto here to have the machine say
"this is what I got" and compare that to what turned up in the struct (in
case there is a mismatch between the correct proto and an incorrect
deserializer)

On Mon, Apr 11, 2016 at 3:22 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/api/serialization_test.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r59162670>
> :
>
> > @@ -58,6 +61,15 @@ func fuzzInternalObject(t *testing.T, forVersion unversioned.GroupVersion, item
> >  	return item
> >  }
> >
> > +func dataAsString(data []byte) string {
> > +	dataString := string(data)
> > +	if !strings.HasPrefix(dataString, "{") {
> > +		dataString = "\n" + hex.Dump(data)
> > +		proto.NewBuffer(make([]byte, 0, 1024)).DebugPrint("decoded object", data)
>
> Just curious - what's that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/e6879a33f46577ce93328230a28453244c3a408f#r59162670>
>

wojtek-t commented on pkg/util/framer/framer.go:
```diff
@@ -0,0 +1,156 @@
+
+type lengthDelimitedFrameReader struct {
+	r         io.Reader
+	remaining int
```
Can you please add a comment about semantics of "remaining"?

wojtek-t commented on pkg/watch/versioned/register.go:
```diff
@@ -0,0 +1,84 @@
+		out.Object.Raw = t.Raw
+	case nil:
+	default:
+		out.Object.Object = in.Object
```
Shouldn't this be a DeepCopy?

smarterclayton replied:
In this case we're calling convert which does not guarantee the incoming
object is mutated.  Since runtime.Object is an interface and we use some
other types in interfaces that aren't purely serializable
(runtime.Encodable) it's not safe to copy.  I can make a comment in the
godoc of runtime.Extension that it does not perform a copy.

On Mon, Apr 11, 2016 at 4:15 AM, Wojciech Tyczynski <
notifications@github.com> wrote:

> In pkg/watch/versioned/register.go
> <https://github.com/kubernetes/kubernetes/pull/23806#discussion_r59167904>
> :
>
> > +		Convert_versioned_Event_to_watch_Event,
> > +		Convert_versioned_InternalEvent_to_versioned_Event,
> > +		Convert_watch_Event_to_versioned_Event,
> > +		Convert_versioned_Event_to_versioned_InternalEvent,
> > +	)
> > +}
> > +
> > +func Convert_watch_Event_to_versioned_Event(in *watch.Event, out *Event, s conversion.Scope) error {
> > +	out.Type = string(in.Type)
> > +	switch t := in.Object.(type) {
> > +	case *runtime.Unknown:
> > +		// TODO: handle other fields on Unknown and detect type
> > +		out.Object.Raw = t.Raw
> > +	case nil:
> > +	default:
> > +		out.Object.Object = in.Object
>
> Shouldn't this be a DeepCopy?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23806/files/e6879a33f46577ce93328230a28453244c3a408f#r59167904>
>

wojtek-t commented on pkg/watch/versioned/register.go:
```diff
@@ -0,0 +1,84 @@
+func Convert_versioned_Event_to_watch_Event(in *Event, out *watch.Event, s conversion.Scope) error {
+	out.Type = watch.EventType(in.Type)
+	if in.Object.Object != nil {
+		out.Object = in.Object.Object
```
Shouldn't this be a DeepCopy?

wojtek-t commented:
@smarterclayton - sorry for long delay in reviewing this
I've added few more minor comments. Also, I've manually merged the "time-proto" PR, so you can remove the first two commits from this PR.

wojtek-t commented:
LGTM - thanks!
